File: Best_Research/#Done/Rust/LearningRust/Units/Units/src/units/force/display.rs
use std::fmt;

use crate::internal::format_unit_value;

use super::quantity::{Force, ForceUnit};

impl Force {
    #[must_use]
    pub const fn display_as(self, unit: ForceUnit) -> ForceDisplay {
        ForceDisplay {
            value: self,
            unit,
            precision: None,
        }
    }

    #[must_use]
    pub const fn display_as_precision(self, unit: ForceUnit, precision: usize) -> ForceDisplay {
        ForceDisplay {
            value: self,
            unit,
            precision: Some(precision),
        }
    }
}

#[derive(Copy, Clone, Debug)]
pub struct ForceDisplay {
    value: Force,
    unit: ForceUnit,
    precision: Option<usize>,
}

impl fmt::Display for ForceDisplay {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        let value = match self.unit {
            ForceUnit::Newtons => self.value.as_newtons(),
            ForceUnit::Millinewtons => self.value.as_millinewtons(),
            ForceUnit::Kilonewtons => self.value.as_kilonewtons(),
        };

        format_unit_value(formatter, value, self.unit.symbol(), self.precision)
    }
}

impl fmt::Display for Force {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.display_as(ForceUnit::Newtons).fmt(formatter)
    }
}

File: Best_Research/#Done/C++/Units-CppXmake/Physics/force/display.cpp
#include "Physics/display/display.hpp"

#include <ostream>

namespace Physics::force
{

    const char* symbol(ForceUnit unit)
    {
        switch (unit)
        {
        case ForceUnit::Newtons:
            return "N";
        case ForceUnit::Millinewtons:
            return "mN";
        case ForceUnit::Kilonewtons:
            return "kN";
        }

        return "";
    }

    double valueInUnit(Force value, ForceUnit unit)
    {
        switch (unit)
        {
        case ForceUnit::Newtons:
            return value.asNewtons();
        case ForceUnit::Millinewtons:
            return value.asMillinewtons();
        case ForceUnit::Kilonewtons:
            return value.asKilonewtons();
        }

        return value.asNewtons();
    }

    std::ostream& operator<<(std::ostream& stream, Force value)
    {
        return stream << value.displayAs(ForceUnit::Newtons);
    }

} // namespace Physics::force
